### PR TITLE
[TRTLLM-15400][perf] fuse per-token AdaLN for VisualGen Wan 2.2 5B

### PR DESCRIPTION
**Pull request by chang-l**
# [TRTLLM-15400][perf] fuse per-token AdaLN for VisualGen Wan 2.2 5B

## Scope

This PR targets the shared D=3072 Wan 2.2 5B transformer:

- Standard `Wan2.2-TI2V-5B`, in both T2V and I2V modes.
- FastWan 2.2 5B, which reuses the same transformer architecture and changes only the denoising recipe.

It does **not** replace the Wan 2.2 A14B D=5120 path. A14B continues to use the existing in-tree fused AdaLN/LayerNorm implementation.

The new kernels require BF16 **AdaLN activations**, but they are not restricted to unquantized models. FP8, FP8 block-scale, and NVFP4 configurations still carry BF16 tensors through AdaLN and quantize at the downstream `Linear` boundary. The new AdaLN kernels therefore apply to those weight-quantized paths as long as the runtime tensors satisfy the BF16 and layout contract. They do not themselves emit FP8/FP4 activations or fuse the downstream activation-quantization step.

## Problem

Wan 2.2 TI2V-5B expands timesteps per patch, producing per-token AdaLN inputs (`temb.ndim == 4`). At the production shape, the baseline expression

```python
scale_shift_table.float() + temb.float()
```

materializes large FP32 `[B, S, D]` modulation tensors across transformer blocks. In the measured B200 workload, the three normalization/modulation sites account for about 10.4% of a denoise step.

The existing in-tree Wan fused AdaLN path is specialized for D=5120 and consumes already materialized modulation tensors. Extending only its dimension check would not remove the D=3072 per-token modulator materialization addressed here.

## Change

- Add SM100 CuTe DSL custom ops for per-token AdaLN and gated-residual AdaLN.
- Consume the raw BF16 timestep chunks plus FP32 `scale_shift_table` rows and form the modulation values inside the kernel.
- Cover the three Wan block normalization boundaries while preserving eager residual-rounding semantics.
- Keep kernel loading, eligibility, tensor contracts, layout preparation, and dispatch behind model-level operations in `utils_wan.py`; `transformer_wan.py` remains expressed in Wan normalization and residual operations.
- Register functional `torch.library` custom ops with fake implementations and fixed-stride requirements for `torch.compile(fullgraph=True)`.
- Fail closed to the existing implementation for unsupported hardware, activation dtype, timestep shape, dimension, device, stride, or alignment.

### Eligibility

The optimized path engages when all of the following hold:

- Wan 2.2 5B per-token timestep path (`temb.ndim == 4`).
- BF16 model activations.
- `D % 256 == 0` and `D <= 8192`; production TI2V-5B uses D=3072.
- Exact SM100 runtime with the required tensor layout and alignment.
- CuTe DSL kernels are available.
- The existing D=5120 fusion is not selected.

Weight quantization is intentionally not an eligibility condition. Runtime activation checks remain authoritative and safely fall back if a particular configuration does not present BF16 tensors.

## Evidence

Measured on B200 with an internal study harness on the TRT-LLM 1.3.0rc24 container baseline. Site benchmarks are interleaved A/B medians of 100 runs; parity is checked against an FP32 oracle.

| Metric | Result |
|---|---|
| Wan2.2-TI2V-5B end-to-end denoise, this PR alone | **-4.32%** (43.6 s to 41.7 s; three prompts, spread < 0.01%) |
| Combined with the Wan MLP GELU-epilogue change | **-6.90%** |
| Visual parity against baseline output | LPIPS 0.041 |
| Engagement | 30 blocks x 3 normalization sites = 90 launches per step |
| Per-op numerical parity | Error matches eager; residual path bit-exact or within 1 BF16 ULP |
| In-model normalization/modulation effect | About 2.97 ms to about 1.1 ms per block |

These performance measurements use the BF16 benchmark configuration. FP8/NVFP4 compatibility follows the same BF16 AdaLN activation contract; this PR does not claim a separately measured quantized-model speedup.

## Test coverage

- `test_pertoken_adaln.py` on B200 covers D=768 and D=3072 numerical parity, norm2/norm3 residual rounding, invalid contracts, and full-graph compilation through both the custom ops and the Wan normalization module.
- The standard Wan 2.2 TI2V-5B B200 T2V/I2V comparison asserts that both fused operators execute on SM100 and then compares decoded output with the Hugging Face reference.
- The existing Wan 2.2 TI2V-5B FP8 combined-optimization test now asserts that the per-token AdaLN adapter remains enabled after an SM100 forward.
- Unsupported runtime conditions retain the existing fallback behavior.
- Local `py_compile`, Ruff, the full file-scoped pre-commit suite, test-list validation, and DCO checks pass. This host has no CUDA device or pytest installation, so B200 runtime tests must run in CI.

## Licensing and provenance

`reduce.py` is a byte-identical vendored copy of SGLang's Apache-2.0 warp/CTA reduction helpers at pinned SHA `e1c4db9621f7c4203ee9becd5d5456d4e6bf54f7`, with only the provenance header added. The row-per-CTA scaffold is derived from SGLang's Apache-2.0 ScaleResidualNormScaleShift kernel at the same revision and is credited in the file header. No dependency is added; the implementation uses the already pinned `nvidia-cutlass-dsl==4.5.0`.

## PR checklist

- [x] PR description clearly explains the Wan 2.2 5B scope and motivation.
- [x] PR follows the TRT-LLM coding guidelines.
- [x] Test cases cover numerical parity, invalid contracts, compilation, standard 5B engagement, and FP8 eligibility.
- [x] No public API changes.
- [x] Vendored/derived code includes Apache-2.0 provenance and SPDX headers.
- [x] No new package dependencies.
- [x] CODEOWNERS update is not required.
- [x] Reviewers include CuTe DSL kernel owners and VisualGen owners.

## Dev Engineer Review

- Added Blackwell CuTe DSL fused per-token AdaLN kernels for plain and residual paths.
- Integrated the kernels into all three Wan normalization sites.
- Added runtime validation, fake implementations, compilation caching, and fallback behavior.
- Limited dispatch to eligible SM100 and BF16 configurations, including quantized Wan 5B paths.
- Preserved existing behavior for unsupported configurations.
- Added vendored warp and CTA reduction helpers with Apache-2.0 metadata.
- Added a `num_warps <= WARP_SIZE` assertion to protect single-warp reductions.
- Reported B200 results show a 4.32% end-to-end denoise improvement, or 6.90% with the Wan MLP GELU-epilogue change.

## QA Engineer Review

- Added tests for numerical parity, residual rounding, invalid contracts, `torch.compile`, SM100 engagement, and Hugging Face output comparison.
- Added coverage for the `num_warps` boundary condition.
- Updated Wan pipeline tests to verify fused operator engagement and per-block enablement on SM100.
- Added the per-token AdaLN test file to `tests/integration/test_lists/test-db/l0_b200.yml`.
- CI coverage exists through the B200 test list.
- Verdict: sufficient.

### CODE REVIEW

File: tensorrt_llm/_torch/cute_dsl_kernels/blackwell/pertoken_adaln/pertoken_adaln.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+    if gate_chunk is not None:
+        _validate_bsd(gate_chunk, B, S, D, "gate_chunk")
+        _validate_row(table_gate, D, "table_gate")
+    _validate_row(weight, D, "weight")
+    _validate_row(bias, D, "bias")
+    if shift_chunk is not None:
+        _validate_bsd(shift_chunk, B, S, D, "shift_chunk")
+        _validate_bsd(scale_chunk, B, S, D, "scale_chunk")
+        _validate_row(table_shift, D, "table_shift")
+        _validate_row(table_scale, D, "table_scale")
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply tensorrt_llm/_torch/cute_dsl_kernels/blackwell/pertoken_adaln/pertoken_adaln.py]: +1

File: tensorrt_llm/_torch/cute_dsl_kernels/blackwell/pertoken_adaln/pertoken_adaln.py [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+    if t.dtype not in _ROW_DTYPES:
+        raise ValueError(f"{name}: unsupported dtype {t.dtype}")
+    if t.shape != (B, S, D):
+        raise ValueError(f"{name}: expected shape {(B, S, D)}, got {tuple(t.shape)}")
+    if t.stride(-1) != 1:
+        raise ValueError(f"{name}: last dim must be contiguous (stride(-1)==1)")
+    for i in range(t.ndim - 1):
+        if t.stride(i) % D != 0:
+            raise ValueError(f"{name}: stride({i})={t.stride(i)} not divisible by D={D}")
+
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply tensorrt_llm/_torch/cute_dsl_kernels/blackwell/pertoken_adaln/pertoken_adaln.py]: +1

File: tensorrt_llm/_torch/cute_dsl_kernels/blackwell/pertoken_adaln/reduce.py
Comment (kris1025) [inline tensorrt_llm/_torch/cute_dsl_kernels/blackwell/pertoken_adaln/reduce.py:1]: Only one warp’s 32 lanes participate. With D≤8192, vec≥8 → num_warps≤32, so you’re OK at the limit; if vec ever drops or D grows, this silently breaks. Prefer an assert num_warps <= WARP_SIZE in PerTokenAdaLN.__init__, or a multi-step CTA reduce.

**Review by yingguo-trt — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]

**Review by chzblych — APPROVED** [review]

**Comment by luyiyun1021** [comment]
The perf result looks good, and the DRAM-traffic reasoning behind it is clear. I'd like to raise a structural point about the kernel family rather than about this implementation.

After this PR, the tree will have three fused AdaLN kernels: `fusedDiTGateResidNormShiftScaleKernel` (LTX-2), `fusedAdaptiveLayerNormKernel` (Wan 2.2 A14B, D=5120), and the new CuTe DSL `fused_pertoken_adaln{,_residual}` (Wan 2.2 TI2V-5B). All three cover the same shape of computation: normalization plus optional gate/residual plus AdaLN modulate over a `[tokens, D]` row, with an optional NVFP4 epilogue.

On the motivation in the description: avoiding the materialization of `scale_shift_table.float() + temb.float()` is indeed not expressible with the A14B kernel, since `modBase = batchIdx * D` fixes the modulator stride and requires a materialized contiguous `[M, D]` buffer. The LTX-2 kernel, however, already runs exactly this pattern in production, and with the same per-token layout this PR uses. `_get_ada_table_ts_pairs` produces `(scale_shift_table[i] fp32 [D], timestep.reshape(B, T, 6, -1)[:, :, i, :] bf16 [B, T, D])` pairs, which is the direct analogue of this PR's `(table_rows[k], temb[:, :, k])`. The op contract states it explicitly: `scale[b, t, d] = scale_table[d] + scale_ts[b, t, d]`. On the C++ side the per-token case falls out of the existing parameterization — `n_rows = ts.numel() / D` gives `tokens_per_batch = 1`, and the modulator stride is taken from `ts.stride(-2)` (`6*D` for the chunk view); the code comment notes it "handles both flat `[B, D]` and `[B, T_t, K, D]` unbind views".

Mapping the three Wan 5B sites onto the LTX-2 variants:

| Wan 5B site | Expression | Closest LTX-2 variant | Actual delta |
|---|---|---|---|
| norm1 | `LN(x) * (1 + tbl+ts) + (tbl+ts)` | `rmsnorm_shift_scale` | LayerNorm vs RMSNorm; fp32 vs bf16 combine |
| norm2 | `x + attn1*(tbl+ts)` then `LN_affine(w, b)` | `gate_resid_rmsnorm` | plus LN affine; LayerNorm; fp32 combine |
| norm3 | `x + attn2` then `LN(x) * (1 + tbl+ts) + (tbl+ts)` | `resid_rmsnorm_shift_scale` (only NUM_OUT=2 instantiated) | needs NUM_OUT=1; LayerNorm; fp32 combine |

So the differences reduce to three axes — norm type (LayerNorm needs one extra mean reduction), LN affine weight/bias, and combine precision (Wan's eager is fp32, LTX's is bf16 `__hadd2` to match its own eager) — plus a D=3072 tile instantiation. Per-token addressing and modulator strides need no change at all. Each of those reads like a template flag of the same kind the kernel already carries.

The practical consequence I'd like us to think about is reuse. When the next DiT model arrives with a slightly different combination — RMSNorm with affine, or per-token modulation with an FP8 epilogue — adding a fourth kernel may be the cheapest local option, since each of the three carries its own eligibility logic, layout contract, tests, and tuning history. There is also some divergence in capability today: this kernel has no NVFP4 epilogue, which both existing ones have and which is relevant for the quantized Wan 5B configurations this PR enables, while the LTX-2 one has TMA and a warp-specialized pipeline that this one doesn't.

So the question for the team: would it make sense to converge on one general fused AdaLN kernel parameterized over {norm type, affine, gate/residual, modulator granularity and stride, combine precision, output count, epilogue}, with per-model code selecting a specialization? I don't think this PR needs to block on that — the perf is real and the fallback path is safe. A follow-up PR seems like the right place. In the meantime, it would help to record in the description why extending the LTX-2 kernel wasn't the chosen path, so the reasoning is available next time this decision comes up.

**Comment by chang-l** [comment]
>would it make sense to converge on one general fused AdaLN kernel parameterized over {norm type, affine, gate/residual, modulator granularity and stride, combine precision, output count, epilogue}, with per-model code selecting a specialization? 

@luyiyun1021  yes, it makes sense. We now have more data points across different AIGV models to identify a shared abstraction. A retrospective and incremental refactor toward a common parameterized kernel would reduce duplication and provide a better foundation moving forward.  Not only adaLN-class of kernels, the `addmm`/`addmm_act` -style fused kernels also appear broadly reusable across VisualGen models (they are less applicable to typical LLM workloads).

File: tensorrt_llm/_torch/cute_dsl_kernels/blackwell/pertoken_adaln/reduce.py
Comment (chang-l) [reply tensorrt_llm/_torch/cute_dsl_kernels/blackwell/pertoken_adaln/reduce.py:1]: Good catch — agreed, and done: added `assert self.num_warps <= WARP_SIZE` in `PerTokenAdaLN.__init__` (e3e0420fce), plus a unit test covering both the failing construction (D=16384 → 64 warps) and the exact-limit one (D=8192, vec=8 → 32 warps, which `cta_reduce_sum` handles correctly).

**Review by kris1025 — APPROVED** [review]
